### ISSUE TITLE
Switching from Debug to Release generates invalid code on Linux x64

### ISSUE THREAD

**Issue by manu-silicon** [dotnet/coreclr#3753]
I've been successful at running CoreCLR in debug mode on Linux Ubuntu 14.04 x64. I wanted to do some benchmarks, so I built the release mode. When using it, it would crash almost immediately in the following native code:

```
frame #0: 0x00007ffff1c471d4 System.Globalization.Native.so`std::less<int>::operator(this=0x000000000068ff7c, __x=0x0068ff8000000020, __y=0x00007fffffffda14)(int const&, int const&) const + 20 at stl_function.h:235
frame #1: 0x00007ffff1c48674 System.Globalization.Native.so`std::_Rb_tree<int, std::pair<int const, UCollator*>, std::_Select1st<std::pair<int const, UCollator*> >, std::less<int>, std::allocator<std::pair<int const, UCollator*> > >::_M_lower_bound(this=0x000000000068ff7c, __x=0x0068ff8000000000, __y=0x000000000068ff84, __k=0x00007fffffffda14) + 84 at stl_tree.h:1141
frame #2: 0x00007ffff1c4873e System.Globalization.Native.so`std::_Rb_tree<int, std::pair<int const, UCollator*>, std::_Select1st<std::pair<int const, UCollator*> >, std::less<int>, std::allocator<std::pair<int const, UCollator*> > >::find(this=0x000000000068ff7c, __k=0x00007fffffffda14) + 78 at stl_tree.h:1792
frame #3: 0x00007ffff1c4701d System.Globalization.Native.so`std::map<int, UCollator*, std::less<int>, std::allocator<std::pair<int const, UCollator*> > >::find(this=0x000000000068ff7c, __x=0x00007fffffffda14) + 29 at stl_map.h:822
frame #4: 0x00007ffff1c45ec4 System.Globalization.Native.so`GetCollatorFromSortHandle(pSortHandle=0x000000000068ff74, options=1, pErr=0x00007fffffffda74) + 148 at collation.cpp:349
frame #5: 0x00007ffff1c45fdd System.Globalization.Native.so`GlobalizationNative_CompareString(pSortHandle=0x000000000068ff74, lpStr1=0x00007fff5801be44, cwStr1Length=1, lpStr2=0x00007fff5801be64, cwStr2Length=1, options=1) + 77 at collation.cpp:379
```

The ```pSortHandle``` argument seems invalid.

Thanks to @kangaroo who had that problem in the past, I managed to solve the issue by redoing the build but this time by removing any previous artifact of my debug builds (i.e. using ```git clean -xfd```).

It would be nice if one could easily switch from debug to release without having to rebuild from scratch.

**Comment by ellismg** [comment]
When you say switch from debug to release, what do you mean?  Can you give me the exact set of steps you're running (after you've cleaned) to reproduce this?

**Comment by manu-silicon** [comment]
Basically, I started with ```./build.sh``` and then ```./build.sh Release```. Then using the released version of corerun on a program that was working would yield this problem.

**Comment by ellismg** [comment]
Can you share the program you're running or give me an idea of what it is?  Are you copying anything into the output folder with corerun, libcoreclr.so and mscorlib.dll (like parts of corefx?)

**Comment by AndyAyersMS** [comment]
I think there is a build system issue if you switch from debug to release in the same tree without cleaning it first. mscorlib.dll is written to a configuration-neutral location and then via postprocessing ends up in in the appropriate Debug/Release directory. So if you build one way and then immediately build the other, the build system thinks the initial mscorlib is up to date and you can get the wrong version propagated.

A simple fix is to pass /p:Configuration=%__BuildType% to the msocorlib msbuild invocation in build.cmd (or equivalent for build.sh)

**Comment by manu-silicon** [comment]
@AndyAyersMS I'll give it a try next time I'll do the switch. Thanks.

**Comment by janvorli** [comment]
The bug is in the dir.props, here is the fix (I'll send a PR tomorrow, it is too late here now. But if anyone wants to fix it in the meanwhile, feel free to do that:
```
@@ -77,8 +77,8 @@
   <!-- Output paths -->
   <PropertyGroup>
     <BaseIntermediateOutputPath Condition="'$(BaseIntermediateOutputPath)' == ''">$(RootBinDir)obj</BaseIntermediateOutputPath>
-    <IntermediateOutputPath Condition="'$(IntermediateOutputPath)' == ''">$(BaseIntermediateOutputPath)\$(BuildOS).$(BuildArch).$(Configuration)</IntermediateO
-    <OutputPath Condition="'$(OutputPath)' == ''">$(BaseIntermediateOutputPath)\$(BuildOS).$(BuildArch).$(Configuration)</OutputPath>
+    <IntermediateOutputPath Condition="'$(IntermediateOutputPath)' == ''">$(BaseIntermediateOutputPath)\$(BuildOS).$(BuildArch).$(BuildType)</IntermediateOutpu
+    <OutputPath Condition="'$(OutputPath)' == ''">$(BaseIntermediateOutputPath)\$(BuildOS).$(BuildArch).$(BuildType)</OutputPath>
     <FinalOutputPath Condition="'$(FinalOutputPath)' == ''">$(BinDir)</FinalOutputPath>
   </PropertyGroup>
```

**Comment by AndyAyersMS** [comment]
Jan K fixed this in #3855. @manu-silicon let us know if this fixes your issue.

**Comment by manu-silicon** [comment]
@AndyAyersMS @jkotas I've just tried switching and it worked without failure. Thanks!
